### PR TITLE
57 search for provider

### PR DESCRIPTION
**Pull request by devstm**
search for provider by name or location 
image

### CODE REVIEW

**Review by abdulrahman-2020 — APPROVED** [review]

**Review by RandSohail — APPROVED** [review]

**Review by AbdallahGot — APPROVED** [review]

File: server/controllers/user/search.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+          raw: true,
+          group: ['reviews.userId'],
+        });
+        return reviewAvg;
+      }),
+    );
+    const usersWithReviews = rows.map((user, index) => ({
+      ...user.dataValues,
+      avgRating: reviews[index][0].avgRating,
+    }));
```
Comment (LinaYahya) [inline server/controllers/user/search.js:68]:
```suggestion

```

File: server/controllers/user/search.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      const subServiceArray = subservice?.split(',');
+      where['$services.id$'] = subServiceArray;
+    }
+    if (service) {
+      where.mainServiceId = +service;
+    }
+    const { count, rows } = await User.findAndCountAll({
+      limit: 6,
+      offset: (page - 1) * 6,
+      where,
```
Comment (LinaYahya) [inline server/controllers/user/search.js:47]:
```suggestion
      where,
      attributes: {
        exclude: ['password', 'createdAt', 'updatedAt'],
      },
```

File: server/controllers/user/search.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      include: [
+        { model: Location },
```
Comment (LinaYahya) [inline server/controllers/user/search.js:49]: Select attribuates you need only for all models
```suggestion
        { model: Location,  attributes: ['city'] },
```

File: server/utils/validation/searchRequestValidation.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+const Joi = require('joi');
+
+const searchRequestValidation = Joi.object({
+  name: Joi.string(),
+  service: Joi.number(),
+  location: Joi.number(),
+  subservice: Joi.string(),
```
Comment (LinaYahya) [inline server/utils/validation/searchRequestValidation.js:7]:
```suggestion
  subservice: Joi.array().items(Joi.number())
```

File: server/controllers/user/search.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  SubServices,
+  Review,
+  MainServices,
+} = require('../../database/models');
+
+const getSearchResult = async (req, res, next) => {
+  try {
+    const {
+      name, location, service, subservice, page = 1,
+    } = await searchRequestValidation.validateAsync(req.query);
```
Comment (LinaYahya) [inline server/controllers/user/search.js:18]: try to validate subservice after split it, So if 
```js
if(req.query.subservice){
// split
}
// then validate
await searchRequestValidation.validateAsync({...req.query, subservice: splittedArr})
```

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by LinaYahya — APPROVED** [review]

**Pull request merged**